syntax = "proto3";

option objc_class_prefix = "HLW";

// The ClientHandler service definition.
service ClientHandler {
  // Lists users
  rpc ListUsers (ListRequest) returns (ServerReply) {}
  // Requests to Login
  rpc Login (LoginRequest) returns (ServerReply) {}
  // Requests to Logout
  rpc Logout (LogoutRequest) returns (ServerReply) {}
  // Attempts to Send message
  rpc Send (SendRequest) returns (ServerReply) {}
  // Attempts to receive new messages
  rpc GetMessages (GetRequest) returns (ServerReply) {}
  // Attempts to delete account
  rpc Delete (DeleteRequest) returns (ServerReply) {}
  // Switch to new server upon failure
  rpc SwitchServer (SwitchRequest) returns (ServerReply) {}
}

message LoginRequest {
  string user = 1;
}

message LogoutRequest{
    string user = 1;
}

message ListRequest {
    string args = 1;
}

message SendRequest {
  string user = 1;
  string message = 2;
  string target = 3;
}

message DeleteRequest {
    string user = 1;
}

message GetRequest {
  string user = 1;
}

message SwitchRequest {
  string args = 1;
}

message FatalServerError {
  int32 status = 1;
  string errormessage = 2;
  string host = 3;
  string ip = 4;
}

message SwitchReply {
  int32 status = 1;
  string errormessage = 2;
}

message ServerReply {
  oneof reply {
    LoginReply loginreply = 1;
    LogoutReply logoutreply = 2;
    ListReply listreply = 3;
    DeleteReply deletereply = 4;
    SendReply sendreply = 5;
    GetReply getreply = 6;
    SwitchReply switchreply = 7;
    FatalServerError error = 8;
  }
}

message LoginReply {
  int32 status = 1;
  string errormessage = 2;
  string user = 3;
}

message LogoutReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}

message ListReply {
  int32 status = 1;
  string errormessage = 2;
  string wildcard = 3;
  repeated string user = 4;
}

message DeleteReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
}

message SendReply {
    int32 status = 1;
    string errormessage = 2;
    string user = 3;
    string message = 4;
    string target = 5;
}

message UnreadMessage {
  string sender = 1;
  string message = 2;
  string receiver = 3;
}

message GetReply {
  int32 status = 1;
  string errormessage = 2;
  repeated UnreadMessage message = 3;
}